### PR TITLE
add WeaklyUp cluster member node

### PR DESCRIPTION
**Pull request by netcomm**

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #20634

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 1]
```diff
@@ -1111,6 +1115,10 @@ final class Replicator(settings: ReplicatorSettings) extends Actor with ActorLog
     }
   }
 
+  def receiveWeaklyUpMemberUp(m: Member): Unit =
+    if (matchingRole(m) && m.address != selfAddress)
+      weaklyUpNodes += m.address
+
   def receiveMemberUp(m: Member): Unit =
     if (matchingRole(m) && m.address != selfAddress)
       nodes += m.address
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: remove from the `weaklyUpNodes` Set here?

**Comment by patriknw** [comment]
Refs #18399

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 2]
```diff
@@ -1250,7 +1259,7 @@ final class Replicator(settings: ReplicatorSettings) extends Actor with ActorLog
 
   def receiveGetReplicaCount(): Unit = {
     // selfAddress is not included in the set
-    sender() ! ReplicaCount(nodes.size + 1)
+    sender() ! ReplicaCount(nodes.size + weaklyUpNodes.size + 1)
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: perhaps we should include two counts in the `ReplicaCount`, depending on what we decide for the direct replication

**Comment by patriknw** [comment]
OK TO TEST

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala (hunk 1 above)
Comment (netcomm) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: i miss the path. it is right, when the weakly node become normal Up node, it remove from the weaklyUpNodes. i'll fix it.

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala (hunk 2 above)
Comment (netcomm) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: ‘so if the initial reason for weakly nodes is to use them as soon as possible', if the precondition is established, then maybe the one counts is ok.

Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: We need to decide. I'm on the fence but think the safest approach is to only include the WeaklyUp in the gossip and not include them in the direct replication. That is what you have implemented apart from that then the WeaklyUp should not be included in this count.

Comment (netcomm) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: i agree with you, maybe it's the safest approach.

**Comment by drewhk** [comment]
I agree with not using it for direct writing. The whole concept of "majority" is already tricky in case of dynamic membership, at least convergence has some guarantees over what is considered the current membership.

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala (hunk 2 above)
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: we decided to not use them in the direct replication, i.e. should not be included here either

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala (hunk 1 above)
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: ping

**Comment by patriknw** [comment]
@netcomm looks like this is almost done, would you be able to fix the minor things so that we can merge it?

**Comment by netcomm** [comment]
sorry to see later, i have pull it just now.

**Comment by patriknw** [comment]
LGTM

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 3]
```diff
@@ -1111,15 +1117,22 @@ final class Replicator(settings: ReplicatorSettings) extends Actor with ActorLog
     }
   }
 
-  def receiveMemberUp(m: Member): Unit =
+  def receiveWeaklyUpMemberUp(m: Member): Unit =
     if (matchingRole(m) && m.address != selfAddress)
+      weaklyUpNodes += m.address
```
Comment (ktoso) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: I vaguely remember `union` be faster than `+=` though not sure if that's the case still.
Does not matter that much here though, so don't worry about it :)

**Comment by ktoso** [comment]
LGTM, addition makes a lot of sense, thanks for the contrib @netcomm !

**Comment by netcomm** [comment]
small feature, ^_^

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala (hunk 3 above)
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: that's for `++`, not when adding a single element

Comment (ktoso) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: Ah, thanks for the reminder :)

**Pull request merged**